[CDAP-14282] Adds UI testing infrastructure & initial tests

tonybach opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-14282
UI builds: https://builds.cask.co/browse/CDAP-UDUT145
IT builds: https://builds.cask.co/browse/IT-UPD20

To run tests locally:
- all tests in parallel: `yarn test`.
- snapshot/integration tests only: `yarn jest`.
- run end-to-end (Cypress) tests only: `yarn cypress`.
- run end-to-end (Cypress) tests, with steps rendered in browser: `yarn cypress-open`, then click `Run all specs`

tonybach committed:
[CDAP-14282] Adds UI testing infrastructure & initial tests
- Adds react-testing-library
- Installs necessary packages, and have first snapshot file
- Writes sample react-testing-library test
- Adds first e2e test
- Nukes all exisiting tests, except tests for helper functions
- Adds UI test step to pom.xml
- Adds license headers to tests, deletes unnecessary files, and excludes snapshots from rat check
- Only tries to login when security is enabled
- Adds auth token on secure clusters, and makes sure system artifacts are loaded before running pipeline tests

elfenheart commented on cdap-ui/app/cdap/components/Tags/__tests__/Tags.test.js:
```diff
@@ -0,0 +1,41 @@
+    const inputElem = getByTestId('tag-input');
+    expect(inputElem).toHaveTextContent('');
+    fireEvent.change(inputElem, {target: {value: TEST_TAG}});
+    Mousetrap.trigger('return'); // enter value, and then press Enter to add tag
```
is it possible to just use `fireEvent` for this instead of using Mousetrap?

tonybach replied:
Yes we should be able to. Will update the PR

ajainarayanan reviewed (commented):
@tonybach Overall the changes look good. Couple of things,

1. Could we have fixed version of libraries that we are adding
2. In the cypress test what happens if the NUX is enabled?
3. Referred to the design doc but I couldn't find a clear definition of when to add a snapshot test to which component. Is the assumption of adding a snapshot test is only for dumb components like buttons or popovers etc.,? Was there any specific reason for adding snapshot tests for Tags?
4. Does this PR also address running cypress tests in Bamboo or is it only unit and integration tests as of now?

tonybach commented:
@ajainarayanan 
1. Yes have added fixed versions.
2. Since in the tests so far we're not explicitly testing the behavior of the NUX, before visiting the homepage I set `showWelcome` to `false` in `sessionStorage`. This will hide the NUX: https://github.com/caskdata/cdap/pull/10817/files#diff-a383fc4a0f57f2016c83c830d81af0a5R71
3. We should have a snapshot test for every single React component, so I added the snapshot tests to Tags as well.
4. I set up integration with Bamboo for both:
- For unit/integration tests, I added a step to the pom.xml, so that the tests are always run during the UI builds (UDUT/URUT).
- For e2e/cypress tests, I set up this build https://builds.cask.co/browse/IT-UPD which should run the tests every night, based off whatever is in the `develop` branch.

elfenheart commented on cdap-ui/cypress/integration/pipeline.spec.js:
```diff
@@ -0,0 +1,185 @@
+    cy.get('[data-testid=close-modeless]').click();
+
+    // Delete the pipeline to clean up
+    cy.get('.pipeline-actions-popper').click();
```
if the intent here is to clean up, i think there should be a method `afterEach` that will do the cleanup process?

If the intent is to test the deletion process, then it should be under a different test case, I think.

tonybach replied:
You're right, makes sense to do that. I've moved the cleanup block to be in the `afterEach` hook.

elfenheart commented on cdap-ui/cypress/integration/pipeline.spec.js:
```diff
@@ -0,0 +1,186 @@
+      .click();
+
+    // Assert pipeline no longer exists
+    cy.contains('a', TEST_PIPELINE_NAME, { timeout: 10000 }).should('not.exist');
```
should remove this... there's shouldn't be an assertion on a setup/cleanup process

elfenheart reviewed (dismissed):
👍 LGTM

tonybach commented:
Latest builds are green: https://builds.cask.co/browse/CDAP-UDUT145-14, https://builds.cask.co/browse/IT-UPD20-66.
